internal/errors/timeout.go

```diff
@@ -33,25 +33,78 @@ type TimeoutError interface {
 	timeoutError()
 }
 
-type timeoutError struct {
+// handlerTimeoutError represents a timeout on the handler side.
+type handlerTimeoutError struct {
+	Caller    string
 	Service   string
 	Procedure string
 	Duration  time.Duration
 }
 
-// NewTimeoutError constructs an instance of a TimeoutError for the given
-// service, procedure, and duration waited.
-func NewTimeoutError(Service string, Procedure string, Duration time.Duration) TimeoutError {
-	return timeoutError{
+var _ TimeoutError = handlerTimeoutError{}
+var _ HandlerError = handlerTimeoutError{}
+
+// HandlerTimeoutError constructs an instance of a TimeoutError representing
+// a timeout that occurred during the handler execution, with the caller,
+// service, procedure and duration waited.
+func HandlerTimeoutError(Caller string, Service string, Procedure string, Duration time.Duration) TimeoutError {
+	return handlerTimeoutError{
+		Caller:    Caller,
+		Service:   Service,
+		Procedure: Procedure,
+		Duration:  Duration,
+	}
+}
+
+func (handlerTimeoutError) timeoutError() {}
+func (handlerTimeoutError) handlerError() {}
+
+func (e handlerTimeoutError) Error() string {
+	return fmt.Sprintf(`handler timeout for procedure %q of service %q from caller %q after %v`,
+		e.Procedure, e.Service, e.Caller, e.Duration)
+}
+
+type remoteTimeoutError string
+
+var _ TimeoutError = remoteTimeoutError("")
+
+// RemoteTimeoutError builds a new TimeoutError with the given message.
+//
+// It represents a TimeoutError from a remote handler.
+func RemoteTimeoutError(message string) TimeoutError {
```

Comment:
FYI if your usage is,

    type foo Bar

    func Foo(b Bar) { return foo(b) }

You can do,

    type Foo Bar

Usage will still be `Foo(bar)`.

And since this is an internal API, we don't have to worry about users relying on it always being an alias for `string`.